Add textual inversion XL for Gaudi

dsocek opened the pull request:
# What does this PR do?

This PR adds a script to run Textual Inversion fine-tuning on Gaudi for XL diffusion models, such as "stabilityai/stable-diffusion-xl-base-1.0", "cagliostrolab/animagine-xl-3.1", and so on.  README doc is updated accordingly.

atakaha reviewed (commented):
do we need this arg?

dsocek commented:
> do we need --allow_tf32?

No, I removed in the PR amendment

emascarenhas commented:
@dsocek , Is this still good to review/merge or does it require a sync, make style, fast_tests check?

dsocek commented:
> @dsocek , Is this still good to review/merge or does it require a sync, make style, fast_tests check?

@emascarenhas this PR is was good shape, but I need to add tests, rebase and test on 1.16/1.17.   Also I would need to apply fix from #1134.  I can work on this today.

dsocek commented:
@emascarenhas turns out guidance is not working properly even in the original diffusers.  This PR will need to be slightly reworked, but good news is I managed find a fix.  Created HF Diffusers PR here:  https://github.com/huggingface/diffusers/pull/9010

dsocek commented:
@regisss my equivalent PR fix is now merged in main diffusers branch (https://github.com/huggingface/diffusers/pull/9010).  May be good to also update optimum-habana with the fix too.  Could you please review this PR when you get a chance?

imangohari1 commented:
@dsocek 
Some initial thoughts here:
- Could you rebase/sync on top of OH main?
- Make sure `make style` is done. 
- Could you share the results of `test_diffusers.py` after this changes? I think that won't change since this is a new code. 
- We probably need to add a test for this in `test_diffusers.py`. If so, I could help with that if needed.

imangohari1 commented on examples/stable-diffusion/training/README.md:
````diff
@@ -61,9 +61,94 @@ python textual_inversion.py \
+Assuming the afforemenioned cat toy dataset has been obtained, we can launch textual inversion XL training using:
+
+```bash
+python textual_inversion_sdxl.py \
````
I am getting this error running this. thoughts?
```bash 
{'dynamic_thresholding_ratio', 'thresholding', 'variance_type', 'clip_sample_range', 'rescale_betas_zero_snr'} was not found in config. Values will be initialized to default values.
08/20/2024 23:45:37 - INFO - __main__ - ***** Running training *****
08/20/2024 23:45:37 - INFO - __main__ -   Num examples = 700
08/20/2024 23:45:37 - INFO - __main__ -   Num Epochs = 3
08/20/2024 23:45:37 - INFO - __main__ -   Instantaneous batch size per device = 1
08/20/2024 23:45:37 - INFO - __main__ -   Total train batch size (w. parallel, distributed & accumulation) = 4
08/20/2024 23:45:37 - INFO - __main__ -   Gradient Accumulation steps = 4
08/20/2024 23:45:37 - INFO - __main__ -   Total optimization steps = 500
Steps:   0%|                                                                                                                                         | 0/500 [00:00<?, ?it/s]/usr/local/lib/python3.10/dist-packages/diffusers/models/unets/unet_2d_blocks.py:1369: FutureWarning: `scale` is deprecated and will be removed in version 1.0.0. The `scale` argument is deprecated and will be ignored. Please remove it, as passing it will raise an error in the future. `scale` should directly be passed while calling the underlying pipeline component i.e., via `cross_attention_kwargs`.
  deprecate("scale", "1.0.0", deprecation_message)
/usr/local/lib/python3.10/dist-packages/diffusers/models/unets/unet_2d_blocks.py:2628: FutureWarning: `scale` is deprecated and will be removed in version 1.0.0. The `scale` argument is deprecated and will be ignored. Please remove it, as passing it will raise an error in the future. `scale` should directly be passed while calling the underlying pipeline component i.e., via `cross_attention_kwargs`.
  deprecate("scale", "1.0.0", deprecation_message)
Steps:   1%|▊                                                                                                        | 4/500 [03:40<4:39:55, 33.86s/it, loss=0.287, lr=0.002]Traceback (most recent call last):
  File "/devops/sgohari/tests/codes/pr-reviews/868/optimum-habana/examples/stable-diffusion/training/textual_inversion_sdxl.py", line 1076, in <module>
    main()
  File "/devops/sgohari/tests/codes/pr-reviews/868/optimum-habana/examples/stable-diffusion/training/textual_inversion_sdxl.py", line 851, in main
    for step, batch in enumerate(train_dataloader):
  File "/usr/local/lib/python3.10/dist-packages/accelerate/data_loader.py", line 464, in __iter__
    next_batch = next(dataloader_iter)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 631, in __next__
    data = self._next_data()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 675, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/devops/sgohari/tests/codes/pr-reviews/868/optimum-habana/examples/stable-diffusion/training/textual_inversion_sdxl.py", line 519, in __getitem__
    image = Image.open(self.image_paths[i % self.num_images])
  File "/usr/local/lib/python3.10/dist-packages/PIL/Image.py", line 3431, in open
    fp = builtins.open(filename, "rb")
IsADirectoryError: [Errno 21] Is a directory: '/devops/sgohari/tests/codes/pr-reviews/868/optimum-habana/examples/stable-diffusion/training/cat/.cache'
Steps:   1%|▊                                                                                                        | 4/500 [03:40<7:35:26, 55.09s/it, loss=0.287, lr=0.002]
```

imangohari1 replied:
I had removed this folder and redownloed but got the same error. 
Not sure what happened but removing that .cache folder, worked. thanks.

imangohari1 reviewed (commented):
I am not sure why this is crashing on my end Daniel.

imangohari1 commented on examples/stable-diffusion/training/README.md:
```diff
@@ -61,9 +61,94 @@ python textual_inversion.py \
+image.save(f"cat-backpack_p2.png")
+
+prompt_2 = "A <cat-toy> colored backpack"
+image = pipe(prompt=prompt_1, prompt_2=prompt_2, num_inference_steps=50, guidance_scale=7.5).images[0]
```
```suggestion
image = pipe(prompt=prompt, prompt_2=prompt_2, num_inference_steps=50, guidance_scale=7.5).images[0]
```

or line 141 `prompt_1 = "A <cat-toy> backpack"`

imangohari1 reviewed (changes requested):
@dsocek 
Thanks for working on this on both HF and OH. I've tested this pr and seems working. 
I added some CI tests in attached patch. 
I've tested it on both G2 and G1 and it passed. 
Please review and apply with `git am < 0001*`. 

Will take a final look afterwards. 
Thank you. 
[0001-fea-ci-Added-text-inversion-sdxl.-minor-fix.patch](https://github.com/user-attachments/files/16700877/0001-fea-ci-Added-text-inversion-sdxl.-minor-fix.patch)

dsocek commented:
> @dsocek Thanks for working on this on both HF and OH. I've tested this pr and seems working. I added some CI tests in attached patch. I've tested it on both G2 and G1 and it passed. Please review and apply with `git am < 0001*`.
> 
> Will take a final look afterwards. Thank you. [0001-fea-ci-Added-text-inversion-sdxl.-minor-fix.patch](https://github.com/user-attachments/files/16700877/0001-fea-ci-Added-text-inversion-sdxl.-minor-fix.patch)

Thanks @imangohari1, I added your CI test implementation and also rebased to latest main

imangohari1 reviewed (approved):
LGTM! 
@regisss Could you pls. do a final review on this PR. it's been on a queue for a while. 
Thank you.

emascarenhas commented:
Please sync your PR with main/upstream and fix any merge conflicts. Thank you.

emascarenhas commented:
> > Please sync your PR with main/upstream and fix any merge conflicts. Thank you.
> 
> @emascarenhas rebased and resolved conflicts.
> 
> @imangohari1 this good PR has been neglected for far too long.. Lets merge!

Thanks @dsocek , I'm adding it to our queue to review. Can you confirm you also ran fast tests and relevant slow tests and those did not generate new failures.

dsocek commented:
@emascarenhas fast tests and relevant slow test **PASS** with the current PR:

```bash
# Setup
make test_installs
export GAUDI2_CI=1

# Run fast tests
export RUN_SLOW=false
python -m pytest tests/test_diffusers.py

# Output of fast tests
... ============== 112 passed, 48 skipped, 280 warnings in 1277.10s (0:21:17)

# Run relevant slow tests
export RUN_SLOW=true
python -m pytest tests/test_diffusers.py -v -s -k "test_textual_inversion_sdxl"

# Output of slow tests
... ============== 1 passed, 159 deselected, 5 warnings in 293.07s (0:04:53) 
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_868). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
